File: frontend/src/services/restaurantService.ts
import { api } from './api';

export interface Restaurant {
  id: string;
  name: string;
  description: string;
  phone: string;
  email: string;
  imageUrl: string;
  cuisineTypes: string[];
  isActive: boolean;
  isAcceptingOrders: boolean;
  preparationTimeMinutes: number;
  minimumOrderAmount: number;
  deliveryFee: number;
  deliveryRadiusKm: number;
  rating: number;
  totalReviews: number;
  latitude: number;
  longitude: number;
  openingTime: string;
  closingTime: string;
}

export interface MenuItem {
  id: string;
  name: string;
  description: string;
  price: number;
  originalPrice: number;
  imageUrl: string;
  isVegetarian: boolean;
  isVegan: boolean;
  isGlutenFree: boolean;
  spiceLevel: number;
  preparationTimeMinutes: number;
  isAvailable: boolean;
  isFeatured: boolean;
  displayOrder: number;
  tags: string;
}

export interface MenuCategory {
  id: string;
  name: string;
  description: string;
  displayOrder: number;
  imageUrl: string;
  isActive: boolean;
  items: MenuItem[];
}

export interface PageResponse<T> {
  content: T[];
  totalPages: number;
  totalElements: number;
  number: number;
  size: number;
}

export const restaurantService = {
  async getAll(params?: { page?: number; size?: number }) {
    return api.get<PageResponse<Restaurant>>('/restaurants', { params });
  },

  async getAllList() {
    return api.get<Restaurant[]>('/restaurants/list');
  },

  async getById(id: string) {
    return api.get<Restaurant>(`/restaurants/${id}`);
  },

  async searchNearby(params: { lat: number; lon: number; radius?: number }) {
    return api.get<Restaurant[]>('/restaurants/search', { params });
  },

  async getMyRestaurants() {
    return api.get<Restaurant[]>('/restaurants/my');
  },

  async create(data: Partial<Restaurant>) {
    return api.post<Restaurant>('/restaurants', data);
  },

  async update(id: string, data: Partial<Restaurant>) {
    return api.put<Restaurant>(`/restaurants/${id}`, data);
  },

  async delete(id: string) {
    return api.delete(`/restaurants/${id}`);
  },

  async toggleActive(id: string, active: boolean) {
    return api.patch<Restaurant>(`/restaurants/${id}/active`, null, {
      params: { active },
    });
  },

  async toggleAcceptingOrders(id: string, accepting: boolean) {
    return api.patch<Restaurant>(`/restaurants/${id}/accepting-orders`, null, {
      params: { accepting },
    });
  },

  async getMenu(restaurantId: string) {
    return api.get<MenuCategory[]>(`/restaurants/${restaurantId}/menu/categories/with-items`);
  },

  async getCategories(restaurantId: string) {
    return api.get<MenuCategory[]>(`/restaurants/${restaurantId}/menu/categories`);
  },

  async getItems(restaurantId: string, categoryId?: string) {
    return api.get<MenuItem[]>(`/restaurants/${restaurantId}/menu/items`, {
      params: categoryId ? { categoryId } : {},
    });
  },
};

File: backend/src/main/java/com/mealmesh/restaurant/service/RestaurantService.java
package com.mealmesh.restaurant.service;

import com.mealmesh.common.exception.BadRequestException;
import com.mealmesh.common.exception.ResourceNotFoundException;
import com.mealmesh.restaurant.dto.RestaurantRequest;
import com.mealmesh.restaurant.dto.RestaurantResponse;
import com.mealmesh.restaurant.entity.Restaurant;
import com.mealmesh.restaurant.entity.RestaurantAddress;
import com.mealmesh.restaurant.repository.RestaurantRepository;
import com.mealmesh.restaurant.repository.RestaurantAddressRepository;
import com.mealmesh.user.entity.User;
import com.mealmesh.user.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Slf4j
public class RestaurantService {

    private final RestaurantRepository restaurantRepository;
    private final RestaurantAddressRepository addressRepository;
    private final UserRepository userRepository;

    private User getUser(UUID userId) {
        return userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));
    }

    @Transactional
    public RestaurantResponse create(RestaurantRequest request, UUID ownerId) {
        User owner = getUser(ownerId);

        if (restaurantRepository.existsByOwnerAndName(owner, request.getName())) {
            throw new BadRequestException("Restaurant with this name already exists");
        }

        Restaurant restaurant = Restaurant.builder()
                .owner(owner)
                .name(request.getName())
                .description(request.getDescription())
                .phone(request.getPhone())
                .email(request.getEmail())
                .imageUrl(request.getImageUrl())
                .cuisineTypes(serializeCuisineTypes(request.getCuisineTypes()))
                .preparationTimeMinutes(request.getPreparationTimeMinutes() != null ? request.getPreparationTimeMinutes() : 30)
                .minimumOrderAmount(request.getMinimumOrderAmount() != null ? request.getMinimumOrderAmount() : BigDecimal.ZERO)
                .deliveryFee(request.getDeliveryFee() != null ? request.getDeliveryFee() : BigDecimal.ZERO)
                .deliveryRadiusKm(request.getDeliveryRadiusKm() != null ? request.getDeliveryRadiusKm() : new BigDecimal("5.00"))
                .openingTime(request.getOpeningTime() != null ? java.time.LocalTime.parse(request.getOpeningTime()) : null)
                .closingTime(request.getClosingTime() != null ? java.time.LocalTime.parse(request.getClosingTime()) : null)
                .isActive(true)
                .isAcceptingOrders(true)
                .build();

        restaurant = restaurantRepository.save(restaurant);

        if (request.getAddress() != null) {
            RestaurantAddress address = RestaurantAddress.builder()
                    .restaurant(restaurant)
                    .streetAddress(request.getAddress().getStreetAddress())
                    .city(request.getAddress().getCity())
                    .state(request.getAddress().getState())
                    .postalCode(request.getAddress().getPostalCode())
                    .country(request.getAddress().getCountry())
                    .latitude(request.getAddress().getLatitude())
                    .longitude(request.getAddress().getLongitude())
                    .isDefault(request.getAddress().getIsDefault() != null ? request.getAddress().getIsDefault() : true)
                    .build();
            addressRepository.save(address);
            restaurant.setLatitude(address.getLatitude());
            restaurant.setLongitude(address.getLongitude());
        }

        log.info("Restaurant created: {} by owner: {}", restaurant.getName(), owner.getEmail());
        return RestaurantResponse.from(restaurant);
    }

    @Transactional
    @org.springframework.cache.annotation.CacheEvict(value = "restaurants", key = "#id")
    public RestaurantResponse update(UUID id, RestaurantRequest request, UUID ownerId) {
        User owner = getUser(ownerId);
        Restaurant restaurant = restaurantRepository.findByIdAndOwner(id, owner)
                .orElseThrow(() -> new ResourceNotFoundException("Restaurant not found or access denied"));

        restaurant.setName(request.getName());
        restaurant.setDescription(request.getDescription());
        restaurant.setPhone(request.getPhone());
        restaurant.setEmail(request.getEmail());
        restaurant.setImageUrl(request.getImageUrl());
        restaurant.setCuisineTypes(serializeCuisineTypes(request.getCuisineTypes()));
        restaurant.setPreparationTimeMinutes(request.getPreparationTimeMinutes() != null ? request.getPreparationTimeMinutes() : restaurant.getPreparationTimeMinutes());
        restaurant.setMinimumOrderAmount(request.getMinimumOrderAmount() != null ? request.getMinimumOrderAmount() : restaurant.getMinimumOrderAmount());
        restaurant.setDeliveryFee(request.getDeliveryFee() != null ? request.getDeliveryFee() : restaurant.getDeliveryFee());
        restaurant.setDeliveryRadiusKm(request.getDeliveryRadiusKm() != null ? request.getDeliveryRadiusKm() : restaurant.getDeliveryRadiusKm());
        if (request.getOpeningTime() != null) {
            restaurant.setOpeningTime(java.time.LocalTime.parse(request.getOpeningTime()));
        }
        if (request.getClosingTime() != null) {
            restaurant.setClosingTime(java.time.LocalTime.parse(request.getClosingTime()));
        }

        restaurant = restaurantRepository.save(restaurant);
        log.info("Restaurant updated: {}", restaurant.getName());
        return RestaurantResponse.from(restaurant);
    }

    @Transactional(readOnly = true)
    @org.springframework.cache.annotation.Cacheable(value = "restaurants", key = "#id")
    public RestaurantResponse getById(UUID id) {
        Restaurant restaurant = restaurantRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Restaurant not found"));
        return RestaurantResponse.from(restaurant);
    }

    @Transactional(readOnly = true)
    public Page<RestaurantResponse> getAllActive(Pageable pageable) {
        return restaurantRepository.findByIsActiveTrueAndIsAcceptingOrdersTrue(pageable)
                .map(RestaurantResponse::from);
    }

    @Transactional(readOnly = true)
    public List<RestaurantResponse> getAllActive() {
        return restaurantRepository.findByIsActiveTrueAndIsAcceptingOrdersTrue().stream()
                .map(RestaurantResponse::from)
                .toList();
    }

    @Transactional(readOnly = true)
    public List<RestaurantResponse> search(BigDecimal lat, BigDecimal lon, BigDecimal radius) {
        List<Restaurant> restaurants = restaurantRepository.findNearbyRestaurants(lat, lon, radius);
        return restaurants.stream().map(RestaurantResponse::from).toList();
    }

    @Transactional(readOnly = true)
    public List<RestaurantResponse> getByOwner(UUID ownerId) {
        User owner = getUser(ownerId);
        return restaurantRepository.findByOwner(owner).stream()
                .map(RestaurantResponse::from)
                .toList();
    }

    @Transactional
    public void delete(UUID id, UUID ownerId) {
        User owner = getUser(ownerId);
        Restaurant restaurant = restaurantRepository.findByIdAndOwner(id, owner)
                .orElseThrow(() -> new ResourceNotFoundException("Restaurant not found or access denied"));
        restaurantRepository.delete(restaurant);
        log.info("Restaurant deleted: {}", id);
    }

    @Transactional
    public RestaurantResponse toggleActive(UUID id, UUID ownerId, boolean active) {
        User owner = getUser(ownerId);
        Restaurant restaurant = restaurantRepository.findByIdAndOwner(id, owner)
                .orElseThrow(() -> new ResourceNotFoundException("Restaurant not found or access denied"));
        restaurant.setIsActive(active);
        restaurant = restaurantRepository.save(restaurant);
        return RestaurantResponse.from(restaurant);
    }

    @Transactional
    public RestaurantResponse toggleAcceptingOrders(UUID id, UUID ownerId, boolean accepting) {
        User owner = getUser(ownerId);
        Restaurant restaurant = restaurantRepository.findByIdAndOwner(id, owner)
                .orElseThrow(() -> new ResourceNotFoundException("Restaurant not found or access denied"));
        restaurant.setIsAcceptingOrders(accepting);
        restaurant = restaurantRepository.save(restaurant);
        return RestaurantResponse.from(restaurant);
    }

    private String serializeCuisineTypes(List<String> cuisineTypes) {
        if (cuisineTypes == null || cuisineTypes.isEmpty()) {
            return "[]";
        }
        try {
            com.fasterxml.jackson.databind.ObjectMapper mapper = new com.fasterxml.jackson.databind.ObjectMapper();
            return mapper.writeValueAsString(cuisineTypes);
        } catch (Exception e) {
            return "[]";
        }
    }
}